sbnci/Modules/RecoEff/RecoEff_module.cc

```diff
@@ -0,0 +1,417 @@
+      reco_nShowers = n_showers_map[mc_trackID];
+      reco_shower_purity = shower_pur_map[mc_trackID];
+      reco_shower_completeness = shower_comp_map[mc_trackID];
+      reco_track_purity = track_pur_map[mc_trackID];
+      reco_track_completeness = track_comp_map[mc_trackID];
+      reco_track_length = track_length_map[mc_trackID];
+      reco_shower_dEdx = shower_dEdx_map[mc_trackID];
+
+      fParticleTree->Fill();
+    }
+  }
+}
+
+std::vector<art::Ptr<recob::PFParticle> > RecoEff::GetPrimaryPFPs()
+{
+  std::vector<art::Ptr<recob::PFParticle> > primaries;
+  for(unsigned int pfp_i = 0; pfp_i < eHandlePFPs->size(); ++pfp_i) {
+    const art::Ptr<recob::PFParticle> pfp(eHandlePFPs,pfp_i);
+    if(pfp->IsPrimary()){
+      if(pfp->PdgCode() == 12 || pfp->PdgCode() == -12
+         || pfp->PdgCode() == 14 || pfp->PdgCode() == -14){
+        primaries.push_back(pfp);
+      }
+    }
+  }
+  return primaries;
+}
+
+art::Ptr<recob::PFParticle> RecoEff::GetPFP(long unsigned int const &id)
+{
+  const art::Ptr<recob::PFParticle> nullReturn;
+
+  for(unsigned int pfp_i = 0; pfp_i < eHandlePFPs->size(); ++pfp_i) {
+    const art::Ptr<recob::PFParticle> pfp(eHandlePFPs,pfp_i);
+    if(pfp->Self() == id) return pfp;
+  }
+  return nullReturn;
+}
+
+float RecoEff::Purity(std::vector< art::Ptr<recob::Hit> > const &objectHits, int const &trackID)
+{
+  std::map<int,int> objectHitsMap;
+
+  for(unsigned int i = 0; i < objectHits.size(); ++i) {
+    objectHitsMap[TruthMatchUtils::TrueParticleID(clockData,objectHits[i],true)]++;
+  }
+  return objectHitsMap[trackID]/static_cast<float>(objectHits.size());
+}
+
+float RecoEff::Completeness(std::vector< art::Ptr<recob::Hit> > const &objectHits, int const &trackID)
+{
+  std::map<int,int> objectHitsMap;
+
+  for(unsigned int i = 0; i < objectHits.size(); ++i) {
+    objectHitsMap[TruthMatchUtils::TrueParticleID(clockData,objectHits[i],true)]++;
+  }
+  return objectHitsMap[trackID]/static_cast<float>(hits_map[trackID]);
+}
+
+float RecoEff::TrackLength(art::Ptr<recob::Track> const &track)
+{
+  float length = 0;
+  int nTrajPoints = track->NumberTrajectoryPoints();
+
+  if(nTrajPoints < 2) return length;
+
+  for(int point = 1; point < nTrajPoints; ++point) {
+    TVector3 l = track->LocationAtPoint<TVector3>(point);
+    if(l.X() == -999 || l.Y() == -999 || l.Z() == -999) break;
+
+    TVector3 diff = track->LocationAtPoint<TVector3>(point) - track->LocationAtPoint<TVector3>(point-1);
+    length += TMath::Sqrt(diff.Mag2());
+  }
+  return length;
+}
+
+float RecoEff::TrueTrackLength(art::Ptr<simb::MCParticle> const &particle)
+{
+  float length = 0;
+  unsigned int nTrajPoints = particle->NumberTrajectoryPoints();
+
+  if(nTrajPoints < 2) return length;
+
+  for(unsigned int point = 1; point < nTrajPoints; ++point) {
+    TVector3 l = particle->Position(point).Vect();
+    if(l.X() > 200 || l.X() < -200 || l.Y() > 200 || l.Y() < -200 ||
+       l.Z() > 500 || l.Z() < 0) break;
```

Comment:
hard coded values.  Use the geom service?

Reply:
Done [here](https://github.com/SBNSoftware/sbnci/pull/7/commits/ed2c28fd0cd09bef4f5aabe420e12c0a2a1414a5)